contracts/Marketplace.sol

```diff
@@ -92,6 +98,61 @@ contract Marketplace is Collateral, Proofs {
     require(token.transfer(slot.host, amount), "Payment failed");
   }
 
+  /// @notice Withdraws storage request funds back to the client that deposited them.
+  /// @dev Request must be expired, must be in RequestState.New, and the transaction must originate from the depositer address.
+  /// @param requestId the id of the request
+  function withdrawFunds(bytes32 requestId) public marketplaceInvariant {
+    Request memory request = requests[requestId];
+    require(block.timestamp > request.expiry, "Request not yet timed out");
+    require(request.client == msg.sender, "Invalid client address");
+    RequestContext storage context = requestContexts[requestId];
+    require(context.state == RequestState.New, "Invalid state");
+
+    uint256 amount = _price(request);
+    funds.sent += amount;
+    funds.balance -= amount;
+    token.transfer(msg.sender, amount);
```

Comment:
I would move the transfer to the very end of the function, to confirm to the [check-effects-interations pattern](https://docs.soliditylang.org/en/v0.8.16/security-considerations.html#use-the-checks-effects-interactions-pattern). 

We should probably also check the return value of the transfer function:

```suggestion
    require(token.transfer(msg.sender, amount), "Payment failed);
```